airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/source/jdbc/AbstractJdbcSource.kt

```diff
@@ -251,6 +252,45 @@ abstract class AbstractJdbcSource<Datatype>(
         )
     }
 
+    /**
+     * Checks that current user can SELECT from the tables in the schemas. We can override this
+     * function if it takes too long to finish for a particular database source connector.
+     */
+    @Throws(Exception::class)
+    protected open fun checkUserHasPrivileges(config: JsonNode?, database: JdbcDatabase) {
+        var schemas = ArrayList<String>()
+        if (config!!.has(JdbcUtils.SCHEMAS_KEY) && config[JdbcUtils.SCHEMAS_KEY].isArray) {
+            for (schema in config[JdbcUtils.SCHEMAS_KEY]) {
+                schemas.add(schema.asText())
+            }
+        }
+        // if UI has schemas specified, check if the user has select access to any table
+        if (schemas.isNotEmpty()) {
+            for (schema in schemas) {
+                LOGGER.info {
+                    "Checking if the user can perform select to any table in schema: $schema"
+                }
+                val tablesOfSchema = database.metaData.getTables(null, schema, "%", null)
+                if (tablesOfSchema.next()) {
+                    var privileges =
+                        getPrivilegesTableForCurrentUser<JdbcPrivilegeDto>(database, schema)
+                    if (privileges.isEmpty()) {
```

Comment:
Is there a way to look at what privilege is actually granted to the user?
Can there be a case where privileges exist but insufficient?

Reply:
The short answer is yes. 

Currently, this function `getPrivilegesTableForCurrentUser`, is overridden by pg source connector to use a query to check if a user has `SELECT` and `USAGE`.  And the function simply is an no-op at `AbstractJdbcSource.kt`. (Moreover, this function's implementation is missing for other major SQL db connectors such as mssql and MySQL)

I plan to submit another PR to implement this function at Jdbc source level so that all the SQL db sources can leverage this, and remove the overridden implementation from pg source. 

The implementation uses JDBC's function `getTablePrivilege()`, which will return a result set, listing all the privileges a user has on each table of the schemas. With such information, you can basically check against all kinds of privilege requirements. 

Here is an example implementation where I simply check against `SELECT` for the current user:


```
  @Suppress("UNCHECKED_CAST")
    @Throws(SQLException::class)
    override fun <T> getPrivilegesTableForCurrentUser(
        database: JdbcDatabase?,
        schema: String?
    ): Set<T> {
        val rs = database!!.metaData.getTablePrivileges(null, schema, "%")
        val result = mutableSetOf<T>()
        while (rs.next()) {
            val grantee = rs.getString("GRANTEE")
            val privilege = rs.getString("PRIVILEGE")
            if (grantee == database.metaData.userName && privilege == "SELECT") {
                val tableName = rs.getString("TABLE_NAME")
                val schemaName = rs.getString("TABLE_SCHEM")
                result.add(JdbcPrivilegeDto(grantee, tableName, schemaName, privilege) as T)
            }
        }
        return result.toSet()
    }
```